src/coprocessor/gadgets.rs

```diff
@@ -104,50 +114,425 @@ pub fn construct_cons<F: LurkField, CS: ConstraintSystem<F>>(
+            not_dummy,
+            &data_is_empty_str,
+        )?;
+
+        car.implies_ptr_equal(
+            &mut cs.namespace(|| "data is empty str implies car is nil"),
+            &not_dummy_and_data_is_empty_str,
+            &nil,
+        );
+        cdr.implies_ptr_equal(
+            &mut cs.namespace(|| "data is empty str implies cdr is empty str"),
+            &not_dummy_and_data_is_empty_str,
+            &empty_str,
+        );
+    }
+
+    // data is not empty: it's not nil and it's not the empty string
+    let data_is_not_empty = Boolean::and(
+        &mut cs.namespace(|| "data is not empty"),
+        &data_is_nil.not(),
+        &data_is_empty_str.not(),
+    )?;
+
+    {
+        // when data is not empty, we enforce hash equality
+        let not_dumy_and_data_is_not_empty = Boolean::and(
+            &mut cs.namespace(|| "not dummy and data is not empty"),
+            not_dummy,
+            &data_is_not_empty,
+        )?;
+
+        let hash = hash_poseidon(
+            &mut cs.namespace(|| "hash"),
+            vec![
+                car.tag().clone(),
+                car.hash().clone(),
+                cdr.tag().clone(),
+                cdr.hash().clone(),
+            ],
+            store.poseidon_cache.constants.c4(),
+        )?;
+
+        implies_equal(
+            &mut cs.namespace(|| "data is not empty implies hash equality"),
+            &not_dumy_and_data_is_not_empty,
+            data.hash(),
+            &hash,
+        );
+    }
+
+    Ok((car, cdr, data_is_not_empty))
+}
+
+/// Chains `car_cdr` calls `n` times, returning the accumulated `car`s, the final
+/// `cdr` and the (explored) actual length (`<= n`) of the cons-like `data`. For
+/// example, calling `chain_car_cdr` on "ab" with `n = 4` should return the full
+/// actual length `2` of such string. But if `n` is set to `1`, it will return
+/// `1` as the (explored) length.
+///
+/// It can be used to deconstruct cons-lists into their elements or strings into
+/// their characters.
+///
+/// Since it calls `car_cdr` `n` times, the same panicking logic applies.
+///
+/// ```
+/// use bellpepper_core::{boolean::Boolean, test_cs::TestConstraintSystem, ConstraintSystem};
+/// use pasta_curves::Fq;
+///
+/// use lurk::{
+///     circuit::gadgets::pointer::AllocatedPtr,
+///     field::LurkField,
+///     lem::{circuit::GlobalAllocator, pointers::Ptr, store::Store},
+///     coprocessor::gadgets::{a_ptr_as_z_ptr, chain_car_cdr},
+/// };
+///
+/// let mut cs = TestConstraintSystem::new();
+/// let mut g = GlobalAllocator::default();
+/// let store = Store::<Fq>::default();
+/// let nil = store.intern_nil();
+/// let z_nil = store.hash_ptr(&nil);
+/// let empty_str = store.intern_string("");
+/// let z_empty_str = store.hash_ptr(&empty_str);
+/// g.new_consts_from_z_ptr(&mut cs, &z_nil);
+/// g.new_consts_from_z_ptr(&mut cs, &z_empty_str);
+/// let not_dummy = Boolean::Constant(true);
```

Comment:
here the setup of cs, g, and the calls to `new_consts_from_z_ptr` could be hidden.